runtime/array.c

```diff
@@ -469,6 +488,139 @@ CAMLprim value caml_make_local_vect(value len, value init)
+    // [Lambda_to_flambda] since it is in fact ok, if the size is zero.
+    caml_invalid_argument(
+      "%makearray_dynamic: the only array that can be initialized with "
+      "nothing is a zero-length array");
+  } else if (size > Max_array_wosize) {
+    caml_invalid_argument(
+      "%makearray_dynamic: array size too large (> Max_array_wosize)");
+  } else if (is_local) {
+    res = caml_alloc_local_reserved(size, tag, reserved);
+  } else if (size <= Max_young_wosize) {
+    res = caml_alloc_small_with_reserved(size, tag, reserved);
+  } else {
+    res = caml_alloc_shr_reserved(size, tag, reserved);
+  }
+
+  // XXX mshinwell: should we do this?  Unclear if it happens in the cases
+  // for other layouts, we need to check
+  /* Give the GC a chance to run, and run memprof callbacks */
+  if (!is_local) caml_process_pending_actions ();
+
+  return res;
+}
+
+CAMLprim value caml_makearray_dynamic_scannable_unboxed_product(
+  value v_init, value v_is_local, value v_non_unarized_length)
+{
+  // Some of this is similar to [caml_make_vect], above.
+
+  mlsize_t num_initializers = Wosize_val(v_init);
+  int is_local = Bool_val(v_is_local);
+  mlsize_t non_unarized_length = Long_val(v_non_unarized_length);
+
+  value res;
+  mlsize_t size, i;
+
+  // N.B. [v_init] may be on the local stack!
+
+  if (sizeof(uintnat) != sizeof(double)) {
+    // Just make things easy as regards maximum array lengths for now.
+    // This should have been caught in [Lambda_to_flambda].
+    caml_invalid_argument(
+      "%makearray_dynamic: only supported on 64-bit targets "
+        "(this is a compiler bug)");
+  }
+
+  int tag = 0;
+
+  size = non_unarized_length * num_initializers;
+  if (size == 0) {
+    res = Atom(0);
+  } else if (num_initializers < 1) {
+    // This could happen with void layouts.  We don't rule it out in
+    // [Lambda_to_flambda] since it is in fact ok, if the size is zero.
+    caml_invalid_argument(
+      "%makearray_dynamic: the only array that can be initialized with "
+      "nothing is a zero-length array");
+  } else if (size > Max_array_wosize) {
+    caml_invalid_argument(
+      "%makearray_dynamic: array size too large (> Max_array_wosize)");
+  } else if (is_local) {
+    res = caml_alloc_local(size, tag);
+    for (i = 0; i < size; i++) {
+      Field(res, i) = Field(v_init, i % num_initializers);
+    }
+  } else if (size <= Max_young_wosize) {
+    res = caml_alloc_small(size, tag);
+    for (i = 0; i < size; i++) {
+      Field(res, i) = Field(v_init, i % num_initializers);
+    }
+  } else {
+    int move_init_to_major = 0;
+    for (mlsize_t i = 0; i < non_unarized_length; i++) {
+      if (Is_block(Field(v_init, i)) && Is_young(Field(v_init, i))) {
+        move_init_to_major = 1;
+      }
+    }
+    if (move_init_to_major) {
+      /* We don't want to create so many major-to-minor references,
+         so the contents of [v_init] are moved to the major heap by doing
+         a minor GC. */
```

Comment:
Why is this better than adding them to the remembered set with caml_initialize?

Reply:
Not sure, but I'm just following the existing caml_make_vect code.

Reply:
I assume it's because caml_initializing would create a reference for each array index instead of each element field. It sounds much better to explicitly remember each field instead of forcing a minor GC (and domain barrier), though, and same in caml_make_vect

Reply:
Though with safepoints the barrier is never supposed to be expensive, so maybe this is fine. (Not sure what our current plan is regarding them)

Reply:
I've added a CR, we can discuss separately, since this matches existing well-tested code.